[chore] Change requirements to contribute new components (#44453)
open-telemetry/opentelemetry-collector-contrib: docs/new-components.md

# Adding new components

This page explains how to add your own components to an OpenTelemetry Collector and to the
opentelemetry-collector-contrib repository.

The OpenTelemetry Collector has a pluggable architecture that allows you to build your own
[distribution](https://opentelemetry.io/docs/collector/distributions/) with your own [custom
components](https://opentelemetry.io/docs/collector/building/) using the [OpenTelemetry Collector
Builder](https://opentelemetry.io/docs/collector/custom-collector/). You **don't need** to include

> your component in this repository to be able to use or distribute your component: you can just  host

Reviewer: ```suggestion
your component in this repository to be able to use or distribute your component: you can just host
```

## Adding your component to the opentelemetry-collector-contrib repository

You may donate an existing component or propose a whole new one. We recommend that you **start by
building and hosting your component outside of the opentelemetry-collector-contrib repository**.
After you have gotten some real usage you can contribute your component to this repository.

When you are ready to propose adding your component to this repository, [open an
issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=Sponsor+Needed%2Cneeds+triage&projects=&template=new_component.yaml&title=New+component%3A+)
providing the following information:

* Who's the **sponsor** for your component. A sponsor is an approver or maintainer who will be the
  official reviewer of the code and a code owner for the component. You will need to find a sponsor
  for the component in order for it to be accepted. Ideally, the sponsor is from a different company
  than you. If the person proposing the component and the sponsor are
  from the same company, ensure that you have approval on the overall proposal from other approvers

>   and maintainers.

Reviewer: It would be nice to include some information on how to find a sponsor if you're new to the community.

>   and maintainers. You can use the #otel-collector-dev CNCF Slack channel and the Collector SIG meetings for this.

Author: @axw We discussed this on the 2025-11-26 Collector SIG meeting and there seems to be a preference for Christos wording from the people that spoke up. 

Thinking about it further, while your wording is fine by me we would need to be more stringent on codeowner activity. This is something that we want to do (see open-telemetry/opentelemetry-collector/issues/14107) but I guess until we do it Christos exposition on the call convinced me that we should go with his wording.

What do you think?

>   and maintainers. You can use the #otel-collector-dev CNCF Slack channel and the Collector SIG meetings for this.

Author: I tried to word it as "two codeowners plus approver/maintainer which can be the sponsor", PTAL at ff3ebafb16e0ba4fc1d92057bf9373bf625fc90e

- Type `make gencodeowners`. This will trigger the regeneration of the `.github/CODEOWNERS` file. 

When submitting a component to the community, consider breaking it down into separate PRs as follows:

* **First PR** should include the overall structure of the new component:
  * Readme, configuration, and factory implementation usually using the helper
    factory structs.
  * This PR is usually trivial to review, so the size limit does not apply to
    it.
  * The component should use [`In Development` Stability](https://github.com/open-telemetry/opentelemetry-collector#development) in its README.
  * Before submitting a PR, run the following commands from the root of the repository to ensure your new component is meeting the repo linting expectations:
    * `make checkdoc`
    * `make checkmetadata`
    * `make checkapi`
    * `make goporto`
    * `make crosslink`
    * `make gotidy`
    * `make genotelcontribcol`
    * `make genoteltestbedcol`
    * `make generate`
    * `make multimod-verify`
    * `make generate-gh-issue-templates`
    * `make gengithub`
    * `make addlicense`

> * **Second PR** should include the concrete implementation of the component. If the

Reviewer 2: I think the difference to the "unmaintained" process is, that components do have codeowners, but the components lacks of core functionality.

# Adding new components

This page explains how to add your own components to an OpenTelemetry Collector and to the
opentelemetry-collector-contrib repository.

The OpenTelemetry Collector has a pluggable architecture that allows you to build your own
[distribution](https://opentelemetry.io/docs/collector/distributions/) with your own [custom
components](https://opentelemetry.io/docs/collector/building/) using the [OpenTelemetry Collector
Builder](https://opentelemetry.io/docs/collector/custom-collector/). You **don't need** to include
your component in this repository to be able to use or distribute your component: you can just host
it in your own repository as a Go module and [add it to the OpenTelemetry
registry](https://opentelemetry.io/ecosystem/registry/).

> We recommend hosting your component outside of this repository as a first step.

Reviewer 3: ```suggestion
```

Duplicated line or emphasis?

repository, you may choose to donate it following the steps on the next section. If you have
published your component outside of this repository, your activity, community and popularity within
your own repository will help you make the case for the component to be accepted.

## Adding your component to the opentelemetry-collector-contrib repository

You may donate an existing component or propose a whole new one. We recommend that you **start by
building and hosting your component outside of the opentelemetry-collector-contrib repository**.
After you have gotten some real usage you can contribute your component to this repository.

When you are ready to propose adding your component to this repository, [open an
issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=Sponsor+Needed%2Cneeds+triage&projects=&template=new_component.yaml&title=New+component%3A+)
providing the following information:

* Who's the **sponsor** for your component. A sponsor is an approver or maintainer who will be the
  official reviewer of the code and a code owner for the component. You will need to find a sponsor

>   for the component in order for it to be accepted. Ideally, the sponsor is from a different company

Author: Resolving since I removed 'ideally', let's continue the discussion on the other thread

* Who are the **codeowners** for your component. Codeowners are responsible for the component and
  will be pinged for any issues or reviews needed. You need at least two codeowners for your
  component to be accepted, different from the sponsor. We recommend that the codeowners do not work for the same company.
* A **list of other components that cover similar use cases**. These could be components inside of
  this repository or outside of it. If there are competing implementations, invite their maintainers
  to voice their opinion and collaborate on a common implementation.
* Some **information about your component**, such as the reasoning behind it, use-cases, telemetry data
  types supported, and anything else you think is relevant for us to make a decision about accepting
  the component.
* The **configuration options** your component will accept. This will give us a better understanding of
  what it does, and how it may be implemented.

> [!IMPORTANT]  
> Unstable Collector interfaces may undergo breaking changes. Codeowners must be available to update
> or review their components when such changes happen, otherwise the component will be excluded from
> the default builds.

> Maintenance of components is the responsibility of the code owners. Unmaintained components may be

Author: Done!

# Donating new components

This page explains how to add your own components to an OpenTelemetry Collector and how to donate
them to the opentelemetry-collector-contrib repository.

The OpenTelemetry Collector has a pluggable architecture that allows you to build your own
[distribution](https://opentelemetry.io/docs/collector/distributions/) with your own [custom
components](https://opentelemetry.io/docs/collector/building/) using the [OpenTelemetry Collector
Builder](https://opentelemetry.io/docs/collector/custom-collector/). You **don't need** to include
your component in this repository to be able to use or distribute your component: you can just host
it in your own repository as a Go module and [add it to the OpenTelemetry
registry](https://opentelemetry.io/ecosystem/registry/).

To donate your component you need to start by hosting your component outside of this
repository as a first step. This ensures your component develops and gather community feedback.

> ## Hosting your component outside of opentelemetry-collector-contrib

Author: Turned them into subsections

we require that you **build** your component outside of the opentelemetry-collector-contrib
repository. This is also the fastest way to start using your component and to publish it for others to
consume if you want to. 

A component is a Go module (library) built using the
`go.opentelemetry.io/collector` set of libraries. These libraries contain examples (e.g. see the
example on the [`go.opentelemetry.io/collector/exporter` module
documentation](https://pkg.go.dev/go.opentelemetry.io/collector/exporter)). The official
 documentation also has [a section on how to build various kinds of
components](https://opentelemetry.io/docs/collector/building/). You can also use existing
implementations on this repository as a reference. The key criteria to implement a component is to:

* Implement the
  [component.Component](https://pkg.go.dev/go.opentelemetry.io/collector/component#Component)
  interface
* Provide a configuration structure which defines the configuration of the component
* Provide the implementation which performs the component operation

To **use** your component you can [use the OpenTelemetry Collector

> Builder](https://opentelemetry.io/docs/collector/custom-collector/). Even if you don't publish your

Reviewer 4: ```suggestion
To **use** your component you can use the [OpenTelemetry Collector
Builder](https://opentelemetry.io/docs/collector/custom-collector/). Even if you don't publish your
```

module](https://go.dev/doc/modules/publishing). You can publish multiple components from a single
repository by including the path to the component in the tag: for example, the [`filelog`
receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver#file-log-receiver)
v0.139.0 version is [available as a Go
module](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filelogreceiver@v0.139.0)
because we pushed the
[`receiver/filelogreceiver/v0.139.0`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver)
git tag to this repository. We recommend you make a release at least every time there is a breaking
change on any of the Go modules you depend on.

Finally, to **distribute** your component and make sure others can easily discover it, [add it to
the OpenTelemetry registry](https://opentelemetry.io/ecosystem/registry/adding/).

If you think your component fits the requisites to be on the opentelemetry-collector-contrib
repository, you may choose to donate it following the steps on the next section. Your activity,

> community and popularity within your own repository will help you make the case for the component to

Reviewer 4: I'm not following the wording of "community and popularity", should there be a comma after "community" to make it more clear?

be accepted.

## Adding your component to the opentelemetry-collector-contrib repository

After you have gotten some real usage of a component outside of contrib, you can contribute your component to this repository.
When you are ready to propose adding your component to this repository, [open an
issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=Sponsor+Needed%2Cneeds+triage&projects=&template=new_component.yaml&title=Component+donation%3A+)
providing the following information:

* Who's the **sponsor** for your component. A sponsor is an approver or maintainer who will be the
  official reviewer of the code and may also be a code owner. You will need to find a sponsor for
  the component in order for it to be accepted. The sponsor must be from a different company than

>   you. You can use the #otel-collector-dev CNCF Slack channel and the Collector SIG meetings for

Reviewer 4: > if there is (rough) consensus we can change them if we end up with a very different situation to the one we have right now.

That's helpful, thanks for clarifying. I'm not entirely sold on this, but I don't need to block PR. 👍

>   this.

Reviewer 5: ```suggestion
  official reviewer of the code and may also be a code owner. You will need to get consent from this individual to volunteer as a sponsor for your component. The sponsor must be from a different company than
  you. You can use the #otel-collector-dev CNCF Slack channel and the Collector SIG meetings to announce your intention to donate a component and see if any maintainers or approvers would be interested in sponsoring it. Please note that it is not guaranteed you will find a sponsor, as it depends on the availability of approvers and maintainers to assume responsibility for reviewing the component and maintaining it. If you are unable to find a sponsor, you can still continue to use your component as published in your own repository.
```

> * Who are the **codeowners** for your component. Codeowners are responsible for the component and

Reviewer 5: ```suggestion
* The GitHub handles of the **codeowners** for your component. Codeowners are responsible for the component and
```

  will be pinged for any issues or reviews needed. You need at least three codeowners for your
  component to be accepted, one of which must be an approver or maintainer (which can be the
  sponsor). We recommend that the codeowners do not work for the same company.

> * A **list of other components that cover similar use cases**. These could be components inside of

Reviewer 4: In thinking about the general usage and development process, it would probably be good to add an earlier section or comment with a couple thoughts before someone starts writing their own component:
1. Research if this component already exists
2. Research if anyone's used some other implementation, or an existing receiver in a round-about way to achieve the same purpose.

It would save everyone a lot of time and confusion if research is done as early in the process as possible. We could even reference searching the registry, asking around on CNCF slack (eg the otel-collector and otel-collector-dev channels). That way they're prompted to reach out to the community before getting to the point of donating a component (as it may not be as obvious as we'd hope).

This doesn't block this PR from moving forward, just a thought.

  * Update its `metadata.yaml` file.
    * Mark the stability as `alpha`
    * Add `contrib` to the list of distributions
  * Add it to the `cmd/otelcontribcol` binary by updating the `cmd/otelcontribcol/builder-config.yaml` file.
  * Please also run:
    - `make generate`
    - `make genotelcontribcol`
 
  * The component must be enabled only after sufficient testing and only when it meets [`Alpha` stability requirements](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/component-stability.md#alpha).
* Once your component has reached `Alpha` stability, you may also submit a PR to the [OpenTelemetry Collector Releases](https://github.com/open-telemetry/opentelemetry-collector-releases) repository to include your component in future releases of the OpenTelemetry Collector `contrib` distribution.
* Once a new component has been added to the executable:
  * Please add the component
    to the [OpenTelemetry.io registry](https://github.com/open-telemetry/opentelemetry.io#adding-a-project-to-the-opentelemetry-registry).

> [!IMPORTANT]  
> An 'in development' component that does not make progress towards alpha stability it may be

> > considered [unmaintained](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/component-stability.md#unmaintained) and may be removed.

Reviewer 6: ```suggestion
> An 'in development' component that does not make progress towards alpha stability may be
> considered [unmaintained](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/component-stability.md#unmaintained) and eventually removed.
```

stray "it", repeated "may be"
